### PR TITLE
Switch to lsp_utils.GenericClientHandler to support both ST3 and 4

### PR DESCRIPTION
**Pull request by rchl**
This also solves #9 although that was not the main reason. The main reason it to handle both ST3 and ST4 with the same branch.

BTW. Removed the code that moved the server from the old `Cache/` location as it shouldn't be needed anymore.

### CODE REVIEW

**Comment by rchl** [comment]
Also, just noticed that there was a problem related to having separate branches for ST3 and ST4. Given that the version numbers were the same for both ST3 and ST4 variantss, the user would not get updated from ST3 version when upgrading to ST4.

**Comment by rwols** [comment]
I was also making the changes :D But I suppose it's better to merge this one. Up to deathaxe I guess.

```diff
diff --git a/plugin.py b/plugin.py
index 3153b18..417ccd7 100644
--- a/plugin.py
+++ b/plugin.py
@@ -14,31 +14,17 @@ SERVER_SHA256 = "fb38a67211b53c86ee96892a600760b3085e942ceb1c6249e8edc52993304a0
 
 class LemminxPlugin(AbstractPlugin):
 
-    server_dir = ""
-    server_jar = ""
-
     @classmethod
     def name(cls):
         return "lemminx"
 
     @classmethod
-    def configuration(cls):
-        name = "LSP-" + cls.name()
-        base_name = name + ".sublime-settings"
-        file_path = "Packages/{}/{}".format(__package__, base_name)
-        settings = sublime.load_settings(base_name)
-
-        # prepare server command
-        cls.server_dir = os.path.join(cls.storage_path(), __package__)
-        cls.server_jar = os.path.join(cls.server_dir, SERVER_URL.rsplit("/", 1)[1])
-        settings.set("command", ["java", "-jar", cls.server_jar])
+    def server_dir(cls):
+        return os.path.join(cls.storage_path(), __package__)
 
-        # setup scheme cache directory
-        client_settings = settings.get("settings")
-        client_settings["xml.server.workDir"] = cls.server_dir
-        settings.set("settings", client_settings)
-
-        return settings, file_path
+    @classmethod
+    def server_jar(cls):
+        return os.path.join(cls.server_dir(), SERVER_URL.rsplit("/", 1)[1])
 
     @classmethod
     def needs_update_or_installation(cls):
@@ -46,9 +32,9 @@ class LemminxPlugin(AbstractPlugin):
             # move from cache path to package storage
             old_server_dir = os.path.join(sublime.cache_path(), __package__)
             if os.path.isdir(old_server_dir):
-                shutil.move(old_server_dir, cls.server_dir)
+                shutil.move(old_server_dir, cls.server_dir())
             # check hash
-            with open(cls.server_jar, "rb") as stream:
+            with open(cls.server_jar(), "rb") as stream:
                 checksum = hashlib.sha256(stream.read()).hexdigest()
                 return checksum.lower() != SERVER_SHA256.lower()
         except OSError:
@@ -58,20 +44,31 @@ class LemminxPlugin(AbstractPlugin):
     @classmethod
     def install_or_update(cls):
         # download new server binary
-        os.makedirs(cls.server_dir, exist_ok=True)
-        urlretrieve(url=SERVER_URL, filename=cls.server_jar)
+        os.makedirs(cls.server_dir(), exist_ok=True)
+        urlretrieve(url=SERVER_URL, filename=cls.server_jar())
         if cls.needs_update_or_installation():
-            os.remove(cls.server_jar)
+            os.remove(cls.server_jar())
             raise RuntimeError("Error downloading XML server binary!")
 
         # clear old server binaries
-        for file_name in os.listdir(cls.server_dir):
+        for file_name in os.listdir(cls.server_dir()):
             if os.path.splitext(file_name)[1].lower() != ".jar":
                 continue
 
             try:
-                file_path = os.path.join(cls.server_dir, file_name)
-                if not os.path.samefile(file_path, cls.server_jar):
+                file_path = os.path.join(cls.server_dir(), file_name)
+                if not os.path.samefile(file_path, cls.server_jar()):
                     os.remove(file_path)
             except FileNotFoundError:
                 pass
+
+    @classmethod
+    def on_pre_start(cls, window, initiating_view, folders, config):
+        """Called just before starting the subprocess"""
+        # prepare server command
+        config.command = ["java", "-jar", cls.server_jar()]
+
+    def on_settings_changed(self, dotted):
+        """Called just before notifying workspace/didChangeConfiguration"""
+        # setup scheme cache directory
+        dotted.set("xml.server.workDir", self.server_dir())
```

File: LSP-lemminx.sublime-settings [hunk 1]
```diff
@@ -1,8 +1,17 @@
 {
 	"languages": [
 		{
+			// ST4
 			"languageId": "xml",
-			"document_selector": "text.xml"
+			"document_selector": "text.xml",
```
Comment (rwols) [inline LSP-lemminx.sublime-settings]: This can be `"selector": "text.xml"` and outside `"languages"`

**Comment by deathaxe** [comment]
Adding a dependency to run the plugin, make the code more complicated and bloated (my first impression) and using the older `GenericPlugin` instead of the more straight forward `AbstractPlugin`.

What is the improvement of this PR? I see only downsides.

**Comment by rchl** [comment]
It uses the AbstractPlugin interface behind this common ST3 / ST4 interface.

The main benefit is outlined in the first comment but if you are fine with managing two branches then fell free to close it.

And make sure to change the versioning for st3 and 4 releases so that upgrading to ST4 works correctly

**Comment by deathaxe** [comment]
ST3 branch seems to work fine. It won't receive the nice AbstractPlugin API anytime soon, right? So it works as it works, I guess.
Don't see much effort needed except merging the resource branch to both st3 and st4.

**Comment by rchl** [comment]
Personally, I think that juggling between two branches and having to carefully craft a release for both of them is too much but I don't have to do it here so I'm fine with whatever you choose.

**Comment by rchl** [comment]
And BTW. To properly support disabling and enabling the plugin you need to explicitly use `register_plugin` and `unregister_plugin`. That is currently missing in your code (but is handled with this change).

**Comment by deathaxe** [comment]
I removed that recently, as it caused duplicated registrations in the past.

**Comment by deathaxe** [comment]
The only updates ST3 branch sees are the download url and hash. That's quite easy.

**Comment by deathaxe** [comment]
As an embedded devices engineer I like to keep things easy and straight forward instead of wrapping a simple function into dozens of interfaces for whatever unclear reason.

**Comment by deathaxe** [comment]
@rwols Looks as we don't need to setup anything in `configuration()` anymore with your change, right?

Startup command is handled by `on_pre_start` and the workdir modification by `on_settings_changed`.

**Comment by rchl** [comment]
> As an embedded devices engineer I like to keep things easy and straight forward instead of wrapping a simple function into dozens of interfaces for whatever unclear reason.

I'm not gonna convince you otherwise but the reason is very clear I think - having one common-code base that supports both ST3 and ST4.

You get some you lose some with this. Less work to maintain, a bit more abstraction, and more code (although if you combine the code of the ST3 and ST4 branch, it's probably less in total :)).

Some things could be abstracted even more to make the code almost as small as the current one. Would just need to add a new "ServerManagedGenericHandler` or something like that so that the server part is part of the handler itself.

Also, there was some talk about doing some breaking changes in the future which such abstraction can abstract-away from you.

**Comment by rchl** [comment]
> I removed that recently, as it caused duplicated registrations in the past.

That was fixed in the latest LSP release: https://github.com/sublimelsp/LSP/releases/tag/4070-1.2.0

**Comment by rwols** [comment]
> @rwols Looks as we don't need to setup anything in `configuration()` anymore with your change, right?
> 
> Startup command is handled by `on_pre_start` and the workdir modification by `on_settings_changed`.

 Correct. Also, you should probably add back the `register_plugin` and `unregister_plugin` calls in `plugin_loaded`/`plugin_unloaded`.

**Comment by deathaxe** [comment]
Closing in favor of a more direct approach without adding dependencies or confusion about used LSP-API

**Pull request closed without merging**